### ISSUE TITLE
Orbital response for PM localized orbitals does not appear to agree with finite differences

### ISSUE THREAD

**Issue by jonas-greiner** [fishjojo/pyscfad#59]
Hi everyone,

I am trying to compare the orbital response with respect to an external electric field to finite differences. Unfortunately, I am not able to get these to agree:

```
import numpy as np
import jax
from jax import numpy as jnp
from pyscfad import gto, scf, config
from pyscfad.lo import pipek
from pyscfad.ops import stop_grad

jax.config.update("jax_enable_x64", True)
config.update("pyscfad_scf_implicit_diff", True)


# decimal tolerance
TOL = 5

# finite difference shift
FIN_DIFF = 1e-5

# external electric field
e_field = np.zeros(3, dtype=np.float64)

# gauge origin
gauge_origin = np.zeros(3, dtype=np.float64)


def orb_loc(e_field, mol, gauge_origin):
    """
    orbital localization
    """
    # mf calc
    mf = scf.RHF(mol)

    # electronic contribution due to electric field
    with mol.with_common_origin(gauge_origin):
        ao_dip = mol.intor_symmetric("int1e_r", comp=3)
    ext_el = jnp.einsum("x,xij->ij", e_field, ao_dip)

    # modified one-electron hamiltonian
    h1 = mf.get_hcore()
    mf.get_hcore = lambda *args, **kwargs: h1 + ext_el
    mf.kernel()

    # occupied orbitals
    occ_mo = np.where(mf.mo_occ == 2.0)[0]

    # stop tracing mo coefficients as these only serve as initial guess and the
    # localization condition is implicitly differentiated anyways
    mo_coeff_occ = np.asarray(stop_grad(mf.mo_coeff[:, occ_mo]))

    # perform localization
    mlo = pipek.PM(mol, mo_coeff_occ)
    mlo.pop_method = "mulliken"
    mlo.conv_tol = 1e-12
    _ = mlo.kernel()

    # perform Jacobi sweep to ensure stable minimum is found
    mlo = pipek.jacobi_sweep(mlo)

    # construct untraced initial guess
    u0 = mo_coeff_occ.T @ mf.get_ovlp() @ mlo.mo_coeff

    # perform orbital localization which can be implicitly differentiated, the
    # occupied space is now traced again
    mo_coeff = pipek.pm(
        mol,
        mf.mo_coeff[:, occ_mo],
        conv_tol=1e-12,
        pop_method="mulliken",
        init_guess=u0,
    )

    return mo_coeff


# init molecule
mol = gto.Mole(
    verbose=0,
    output=None,
    basis="sto-3g",
    symmetry=True,
    atom="""
        O        0.000000    0.000000    0.117790
        H        0.000000    0.755453   -0.471161
        H        0.000000   -0.755453   -0.471161
    """,
)
mol.build(trace_coords=False, trace_exp=False, trace_ctr_coeff=False)

# AD gradient
ad_grad = jax.jacrev(orb_loc)(e_field, mol, gauge_origin)

# finite differences gradient
fin_diff_grad = np.empty_like(ad_grad)
for i in range(3):
    shift = np.zeros(3, dtype=np.float64)
    shift[i] = FIN_DIFF
    c_p = orb_loc(
        e_field + shift,
        mol,
        gauge_origin,
    )
    c_m = orb_loc(
        e_field - shift,
        mol,
        gauge_origin,
    )

    # ensure numerical orbital gradient is calculated for the same orbitals
    s = c_p.T @ mol.intor_symmetric("int1e_ovlp") @ c_m
    selected_rows = np.full(s.shape[0], -1)
    signs = np.ones(s.shape[0])
    assigned_rows = set()
    assigned_cols = set()
    flat_indices = np.dstack(np.unravel_index(np.argsort(-np.abs(s).ravel()), s.shape))[
        0
    ]
    for row, col in flat_indices:
        if row not in assigned_rows and col not in assigned_cols:
            selected_rows[col] = row
            signs[col] = np.sign(s[col, row])
            assigned_rows.add(row)
            assigned_cols.add(col)
            if len(assigned_rows) == s.shape[0]:
                break
    c_m = signs * c_m[:, selected_rows]

    # ensure correct minima are found with respect to sign changes due to symmetry
    def fun(mo_m, mo_p):
        return jnp.where(
            pipek.cost_function(np.array([]), mol, mo_m[:, np.newaxis])
            - pipek.cost_function(
                np.array([]), mol, (jnp.sign(mo_p * mo_m) * mo_m)[:, np.newaxis]
            )
            < 1e-13,
            jnp.sign(mo_p * mo_m) * mo_m,
            mo_m,
        )
    c_m = jax.vmap(fun, (1, 1), 1)(c_m, c_p)

    # calculate numerical gradient
    grad = (c_p - c_m) / (2 * FIN_DIFF)

    # ensure correct orbital gradients are compared
    euc_dist = np.linalg.norm(
        np.abs(grad[:, :, np.newaxis]) - np.abs(ad_grad[:, np.newaxis, :, i]), axis=0
    )
    selected_rows = np.full(euc_dist.shape[0], -1)
    assigned_rows = set()
    assigned_cols = set()
    flat_indices = np.dstack(
        np.unravel_index(np.argsort(euc_dist.ravel()), euc_dist.shape)
    )[0]
    for row, col in flat_indices:
        if row not in assigned_rows and col not in assigned_cols:
            selected_rows[col] = row
            assigned_rows.add(row)
            assigned_cols.add(col)
            if len(assigned_rows) == euc_dist.shape[0]:
                break
    fin_diff_grad[:, :, i] = grad[:, selected_rows]

print(ad_grad)
print(fin_diff_grad)
print(np.abs(ad_grad) - np.abs(fin_diff_grad))
```

The code is a bit convoluted because I need to ensure that the phases of the orbitals from finite differences agree and that the correct minima are used with respect to sign changes of specific orbitals due to symmetry. Additionally, the gradients for the correct orbitals have to be compared. Nevertheless, the gradients do not appear to agree. While the majority of the elements are the same, some specific elements are completely different. When I try to do the same for IBOs, the differences are even more pronounced. It would be great if somebody has any insight about what else I need to change to get these to agree.

**Comment by jonas-greiner** [comment]
@fishjojo: Do you have any insight about why this might be? It does not appear to be related to symmetry since the localized orbital response also differs for non-symmetric geometries.

**Comment by fishjojo** [comment]
I tried a simpler system, and the result seems fine.
```
import jax
from pyscfad import numpy as np
from pyscfad import gto, scf
from pyscfad.lo import pipek
from pyscfad import config

config.update('pyscfad_scf_implicit_diff', True)

mol = gto.Mole()
mol.atom = '''
He 0 0 0
He 0 0 .74
'''
mol.basis = 'ccpvdz'
mol.verbose = 5
mol.build()

def pmorb(mol, e_field):
    mf = scf.RHF(mol)
    mf.conv_tol = 1e-14

    with mol.with_common_origin(np.zeros(3)):
        ao_dip = mol.intor_symmetric("int1e_r", comp=3)
    ext_el = np.einsum("x,xij->ij", e_field, ao_dip)

    h1 = mf.get_hcore()
    mf.get_hcore = lambda *a: h1 + ext_el
    ehf = mf.kernel()

    orbloc = pipek.pm(mol, mf.mo_coeff[:,mf.mo_occ>0], conv_tol=1e-10)
    return orbloc

e_field = np.zeros(3)
g = jax.jacrev(pmorb, 1)(mol, e_field)

delta = 1e-4
e_p = np.array([0, 0, delta])
e_m = np.array([0, 0,-delta])

orb_p = pmorb(mol, e_p)
orb_m = pmorb(mol, e_m)
g_fd = (orb_p-orb_m) / (2*delta)

print(g_fd)
print((g_fd - g[..., 2]).max())
```
My experience with finite difference is that it is difficult to get the correct result in general, because the localization can be easily trapped in different local minima, and the gauge of the orbitals is hard to fix.
